engine/backends/proverif/proverif_backend.ml

```diff
@@ -114,14 +114,74 @@ module Print = struct
       method ty_int _ = string "bitstring"
 
       method! item' item =
+        let fun_and_reduc name variants =
+          let constructor = List.hd variants in
+          match constructor with
+          | None -> empty
+          | Some constructor ->
+              let field_prefix =
+                if constructor.is_record then string ""
+                else print#concrete_ident name
+              in
+              let fun_args = constructor.arguments in
+              let fun_args_full =
+                separate_map
+                  (comma ^^ break 1)
+                  (fun (x, y, _z) ->
+                    field_prefix ^^ print#concrete_ident x ^^ string ": "
+                    ^^ print#ty_at Param_typ y)
+                  fun_args
+              in
+              let fun_args_names =
+                separate_map
+                  (comma ^^ break 1)
+                  (fst3 >> fun x -> field_prefix ^^ print#concrete_ident x)
+                  fun_args
+              in
+              let fun_args_types =
+                separate_map
+                  (comma ^^ break 1)
+                  (snd3 >> print#ty_at Param_typ)
+                  fun_args
+              in
+              let fun_line =
+                string "fun" ^^ space ^^ print#concrete_ident name
+                ^^ iblock parens fun_args_types
+                ^^ string ": state."
+              in
+              let reduc_line =
+                string "reduc forall " ^^ iblock Fn.id fun_args_full
+                ^^ string ";"
+              in
+              let build_accessor (ident, ty, attr) =
+                string "accessor_" ^^ print#concrete_ident name ^^ string "_"
```

Comment:
Another small, thing: [pprint exposes `underscore`](http://cambium.inria.fr/~fpottier/pprint/doc/pprint/PPrint/index.html#val-underscore), so you can write:
```suggestion
                string "accessor_" ^^ print#concrete_ident name ^^ underscore
```

(same below with `dot`, but that's really not important at all, just a note, you can ignore those things if you like)